### PR TITLE
Add wad plugin

### PR DESCRIPTION
**Pull request by plizonczyk**
This should close https://github.com/andresriancho/w3af/issues/1081

This PR adds new infrastructure plugin, which runs [WAD](https://github.com/CERN-CERT/WAD) tool. It uses [Wappalyzer's](https://github.com/AliasIO/Wappalyzer) rules database to detect technologies used websites, ranging from webservers, through frameworks, to javascript plugins.

As this is my first PR to this project, I'm eager for your extensive code review.

### CODE REVIEW

File: w3af/core/controllers/dependency_check/requirements.py [hunk 1]
```diff
@@ -52,6 +52,7 @@
                      PIPDependency('jinja2', 'Jinja2', '2.7.3'),
                      PIPDependency('vulndb', 'vulndb', '0.0.17'),
                      PIPDependency('markdown', 'markdown', '2.6.1'),
+                     PIPDependency('wad', 'wad', '0.2.0'),
```
Comment (andresriancho) [inline w3af/core/controllers/dependency_check/requirements.py]: https://github.com/andresriancho/w3af-kali/issues/10

**Comment by andresriancho** [comment]
This PR will be rejected since it was made against `master`. Please send it again but against the `develop` branch.

I'll review the code anyways so the next PR is cleaner and merged fast :+1:

File: w3af/plugins/infrastructure/webapp_detection.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+from w3af.core.controllers.plugins.infrastructure_plugin import InfrastructurePlugin
+from w3af.core.controllers.exceptions import RunOnce
+from w3af.core.controllers.misc.decorators import runonce
+from w3af.core.data.kb.info import Info
+from w3af.core.data.options.opt_factory import opt_factory
+from w3af.core.data.options.option_list import OptionList
+
+
+class w3afDetector(Detector):
```
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: Awesome, this is what I meant by subclassing the detector class.

File: w3af/plugins/infrastructure/webapp_detection.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    def get_new_url(self, page):
+        return page.get_url().url_string
+
+    def get_content(self, page, url):
+        return page.get_body()
+
+
+class webapp_detection(InfrastructurePlugin):
+    """
+    Identify technologies used by website based on HTTP response and HTML content
```
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: Please check what happens when you do this:

```
./w3af_console
plugins
infrastructure list
```

This class doc is used in that display. Usually all plugins have one line in the doc which is displayed, not sure if something will break with more lines.

File: w3af/plugins/infrastructure/webapp_detection.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    def discover(self, fuzzable_request):
+        """
+        It calls the "main" from WAD and writes the results to the kb.
+
+        :param fuzzable_request: A fuzzable_request instance that contains
+                                    (among other things) the URL to test.
+        """
+        msg = 'Web application detection (WAD) plugin is starting. ' \
+              'Original WAD authors: CERN-CERT; http://cern.ch/security'
+        om.out.information(msg)
```
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: If all plugins would send a log like this to information() the log would have a lot of garbage. Please remove these two lines.

You can add the original wad authors to the plugin description

File: w3af/plugins/infrastructure/webapp_detection.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+        self._main(fuzzable_request.get_url())
+
+    def _main(self, url):
+        """
+        Based on WAD's main executable
+        """
+
+        detector = w3afDetector(uri_opener=self._uri_opener)
+        results = detector.detect(url, timeout=self._timeout)
```
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: w3af's urlopener already has a default timeout which is set via `http-settings`, I believe there is no value added to the user if we provide a setting to override it in this plugin. Please remove the `self._timeout` and the option to configure it.

File: w3af/plugins/infrastructure/webapp_detection.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    def _report(self, results):
+        """
+        Displays detailed report information to the user and save the data to
+        the kb.
+
+        :return: None.
+        """
+        if results:
+            i = Info('Web applications detection report',
+                     HumanReadableOutput().retrieve(results), 1,
```
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: Please replace the 1 with the real HTTP request ID used to identify this information

File: w3af/plugins/tests/infrastructure/test_webapp_detection.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    _run_configs = {
+        'cfg': {
+            'target': target_url,
+            'plugins': {'infrastructure': (PluginConfig('webapp_detection'),)}
+        }
+    }
+
+    MOCK_RESPONSES = [
+        MockResponse('http://webapp-detection/jquery-in-content', '<script src="jquery.js"></script>'),
+        MockResponse('http://webapp-detection/apache-in-headers-no-content', '', headers={'Server': 'Apache'}),
```
Comment (andresriancho) [inline w3af/plugins/tests/infrastructure/test_webapp_detection.py]: Please respect 80 column limit

File: w3af/plugins/tests/infrastructure/test_webapp_detection.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    def test_apache(self):
+        cfg = self._run_configs['cfg']
+        self._scan(self.target_url + 'apache-in-headers-no-content', cfg['plugins'])
+
+        infos = self.kb.get('webapp_detection', 'webapp_detection')
+
+        self.assertEqual(len(infos), 1, infos)
+        info = infos[0]
+
+        self.assertIn('Apache', info.get_desc())
```
Comment (andresriancho) [inline w3af/plugins/tests/infrastructure/test_webapp_detection.py]: Also assert that jquery is not in desc

File: w3af/plugins/tests/infrastructure/test_webapp_detection.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    def test_jquery(self):
+        cfg = self._run_configs['cfg']
+        self._scan(self.target_url + 'jquery-in-content', cfg['plugins'])
+
+        infos = self.kb.get('webapp_detection', 'webapp_detection')
+
+        self.assertEqual(len(infos), 1, infos)
+        info = infos[0]
+
+        self.assertIn('jQuery', info.get_desc())
```
Comment (andresriancho) [inline w3af/plugins/tests/infrastructure/test_webapp_detection.py]: Also assert that apache is not in desc

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 6 above)
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: This saves the findings to the KB in such a way that it's readable for humans, which is great and will cover 95% of the cases, but something else that we usually do is to `raw_write` data like this in a way that can be easily read/consumed by other plugins.

Not sure how the wad data looks like, but if you could save a dict (or a list, or something similar) containing:
```python
{'jquery': '1.2.3',
'apache': 'unknown'}
```

That would be amazing :)

See: `find . -name '*.py' | xargs grep 'raw_write' | grep -v 'test_'`

**Comment by andresriancho** [comment]
Overall, code and tests look really good, great job!

File: w3af/plugins/infrastructure/webapp_detection.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
+"""
+webapp_detection.py
```
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: The plugin name could be improved, maybe "application_fingerprint" ?

File: w3af/plugins/infrastructure/webapp_detection.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+    def _report(self, results):
+        """
+        Displays detailed report information to the user and save the data to
+        the kb.
+
+        :return: None.
+        """
+        if results:
+            i = Info('Web applications detection report',
```
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: Change the title to `Application fingerprint`

File: w3af/plugins/infrastructure/webapp_detection.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+    :author: Piotr Lizonczyk (piotr.lizonczyk@gmail.com)
+    """
+    def __init__(self):
+        InfrastructurePlugin.__init__(self)
+
+        # User configured
+        self._timeout = 5
+
+    @runonce(exc_class=RunOnce)
```
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: The `RunOnce` is something I would also use, but I was wondering... is this the best we can do? Let me give you a couple of examples:

 * The first fuzzable request that arrives to this plugin is usually http://foo.com/ , in some cases that might just be a redirect to something else like http://foo.com/home/ . w3af doesn't automatically follow redirects, so wad would analyze an empty result

 * http://foo.com/ is just a simple page with links to the different application sections, / doesn't use any JS frameworks nor CMS/etc, but the deeply linked apps do.

In both those cases WAD could find stuff and we're not invoking it to do so.

What if we let w3af call WAD several times until no more new information is collected and then we raise `NoMoreCalls`? Broken pseudo-code:

```python
url_knowledge = self.analyze_url(url)
if self.is_new_knowledge(url_knowledge):
    self.wad_knowledge.merge(url_knowledge)
    self.no_new_knowledge_count = 0
else:
    self.no_new_knowledge_count += 1
    if self.no_new_knowledge_count == 10:
        raise NoMoreCalls
```

I randomly chose `10`, you can change that to any value based on your previous experience with wad

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 3 above)
Comment (plizonczyk) [inline w3af/plugins/infrastructure/webapp_detection.py]: Nothing bad actually happened, lines were concatenated, yet I've changed it.

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 4 above)
Comment (plizonczyk) [inline w3af/plugins/infrastructure/webapp_detection.py]: Sure, I've patterned it on Halberd plugin.

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 5 above)
Comment (plizonczyk) [inline w3af/plugins/infrastructure/webapp_detection.py]: :+1:

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 6 above)
Comment (plizonczyk) [inline w3af/plugins/infrastructure/webapp_detection.py]: ID fixed, I've spotted it hardcoded in mulitple other plugins (I'll later post an issue about that).

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 10 above)
Comment (plizonczyk) [inline w3af/plugins/infrastructure/webapp_detection.py]: Fine.

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 6 above)
Comment (plizonczyk) [inline w3af/plugins/infrastructure/webapp_detection.py]: Raw dict looks like this:
```python
{
    "http://w3af.org": [
        {"type": "web-servers",  "app": "Nginx", "ver": None}, 
        {"type": "cdn", "app": "CloudFlare", "ver": None},
        {"type": "cms,blogs", "app": "WordPress", "ver": None}, 
        {"type": "javascript-frameworks", "app": "jQuery", "ver": None},
        {"type": "photo-galleries,javascript-frameworks", "app": "prettyPhoto", "ver": None}, 
        {"type": "font-scripts", "app": "Font Awesome", "ver": None}, 
        {"type": "font-scripts", "app": "Google Font API", "ver": None}, 
        {"type": "programming-languages", "app": "PHP", "ver": None}  
    ]
}
```

If it's fine, I can easily include it as a switchable option (human/raw data). Do you suggest changing the format to the one that you provided?

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 12 above)
Comment (plizonczyk) [inline w3af/plugins/infrastructure/webapp_detection.py]: Yeah, that's what should be done, I didn't know the mechanism. Should every result be logged in as another Info in knowledge base? If not, then which HTTP response ID should I log?

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 6 above)
Comment (plizonczyk) [inline w3af/plugins/infrastructure/webapp_detection.py]: Alright, I see that it doesn't need to be option, I'll simply store both. :)

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 3 above)
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: :+1:

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 12 above)
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: > Should every result be logged in as another Info in knowledge base?

No, I would log it as one Info with all the merged information. Just write to the `kb` before `raise NoMoreCalls`

> If not, then which HTTP response ID should I log?

You can send a list of IDs in the parameter with all the http responses you've used/merged to create that Info.

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 6 above)
Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: > I've spotted it hardcoded in mulitple other plugins (I'll later post an issue about that).

Thanks!

Comment (andresriancho) [inline w3af/plugins/infrastructure/webapp_detection.py]: > If it's fine, I can easily include it as a switchable option (human/raw data). 

No, you should write to the KB in both formats every time. For reports, GUI, etc. only the Info sub-classes are shown to the user so don't worry about how/if this dict will show to the user.

> Do you suggest changing the format to the one that you provided?

The existing format is OK, keep it please.

**Comment by plizonczyk** [comment]
Closing this MR (bad target branch https://github.com/andresriancho/w3af/pull/12036#issuecomment-129501778)

File: w3af/plugins/infrastructure/webapp_detection.py (hunk 12 above)
Comment (plizonczyk) [inline w3af/plugins/infrastructure/webapp_detection.py]: >>Should every result be logged in as another Info in knowledge base?

>No, I would log it as one Info with all the merged information. Just write to the kb before raise NoMoreCalls

It isn't really possible - we may never reach maximum no_new_knowledge_count. The solution I can come up with is to update the information in KB on every run of the plugin.

**Pull request closed without merging**